fallback_builtins.h

```diff
@@ -4,41 +4,43 @@
 #if defined(_MSC_VER) && !defined(__clang__)
 #if defined(_M_IX86) || defined(_M_AMD64) || defined(_M_IA64) ||  defined(_M_ARM) || defined(_M_ARM64)
 
+#include <assert.h>
 #include <intrin.h>
 #ifdef X86_FEATURES
 #  include "arch/x86/x86_features.h"
 #endif
 
-/* This is not a general purpose replacement for __builtin_ctz. The function expects that value is != 0
- * Because of that assumption trailing_zero is not initialized and the return value of _BitScanForward is not checked
+/* This is not a general purpose replacement for __builtin_ctz. The function expects that value is != 0.
+ * Because of that assumption trailing_zero is not initialized and the return value is not checked.
+ * Tzcnt and bsf give identical results except when input value is 0, therefore this can not be allowed.
+ * If tzcnt instruction is not supported, the cpu will itself execute bsf instead.
+ * Performance tzcnt/bsf is identical on Intel cpu, tzcnt is faster than bsf on AMD cpu.
  */
 static __forceinline int __builtin_ctz(unsigned int value) {
-#ifdef X86_FEATURES
-#  ifndef X86_NOCHECK_TZCNT
-    if (x86_cpu_has_tzcnt)
-#  endif
-        return (int)_tzcnt_u32(value);
-#endif
+    assert(value != 0);
```

Comment:
1. `assert(value != 0);` is needed only for `_BitScanForward` part. `tzcnt` does work for value==0. Same goes for `__builtin_ctzll` bellow.
2. I'm not entirely sure is presence of `_tzcnt_u32` depends on _MSC_VER >= 1700.

Reply:
aren't non-BMI cpus still common? That means that if somebody have 10 year old PC switching to new VS will make your program crash (unless tzcnt is automatically executed as bsf on older CPUs). If it gets executed as bsf, then assert is needed.

Reply:
10+-year old PC is likely running 32-bit Windows Vista by default, and means it can't run latest Visual Studio which is 64-bit binary. I still use Visual Studio 2013 on my Windows Vista laptop (bought in 2009, Intel Core 2 Duo), as it is the most compatible version. There was no need for 64-bit Windows as the maximum installable memory was 4 GB.

Reply:
The assert is for the function as a whole because we define the function as unsafe for zero, so better that we get warnings because they would mean a missing check in the caller code.

I reviewed how some opensource projects used tzcnt, and found this one https://github.com/opencv/opencv/issues/7626 , and that seems to be a pretty clear indication that older MSVC likely did not support this. We could have skipped this backwards compatibility, but I figured it hurts nobody by being there.

Reply:
Also, yes, non-BMI cpus execute tzcnt as bsf because tzcnt gets interpreted as just another bsf instruction on those cpus due to them being identical except for something (bit / flag? I forget the specifics) that bsf just ignores anyway. The incompatibility between them is only for input values of 0.

Reply:
Yeah I believe the opcode is basically the same as BSF.  I tested things with a Core 2 Quad a while back and noticed that they are nearly interchangeable.